### PR TITLE
Use native git variants by default with go-git variants as build tag

### PR DESCRIPTION
**Pull request by zeripath**
This PR provides alternative git functions to go-git derived functions.

The reasoning is that go-git appears to have a number of significant memory issues and in any case large parts of our git infrastructure use git directly in any case.

### CODE REVIEW

**Comment by zeripath** [comment]
Current status of this PR:

* add the tag nogogit to build with non go-git alternatives
* all of the unit tests and integration tests pass with these alternatives.

**Comment by zeripath** [comment]
As an example, prior to this PR running:

```
for i in $(seq 1 100);do curl "http://localhost/gitea/administrator/qtconnectivity/commit/61afd66f148af4a8d3e0cad12b7c9419f1122cbe" &; done
```

on the gitea would balloon in memory use. This no longer happens.

**Comment by lafriks** [comment]
How does this compare if `MADV_DONTNEED` is set? that will be default in golang 1.16 if I'm not mistaken

**Comment by zeripath** [comment]
# Post curl 100 times with nogogit
post-curl-no-go-git

# Post curl 100 times with GODEBUG=MADV_DONTNEED ./gitea web
post-curl-GODEBUG

**Comment by lafriks** [comment]
Can't really read much in these images, does that include memory used by git processes or only gitea? What about performance wise?

**Comment by zeripath** [comment]
This is Gitea's memory. If you zoom in you will see that nogogit Gitea takes up 99.51Mb following the run but gogit Gitea takes up 240Mb. If you run the curl task again gogit will continue to increase.

There's a memory leak somewhere. (And it's in the red tree to the left on the gogit tree)

Peak memory usage for Gitea is higher with go-git than git - although I'm not completely able to assert this.

Now this PR appears slightly slower in the getcommitpaths I will see if I can speed that up though and there's a bug I've noticed elsewhere.

**Comment by lunny** [comment]
How about to move go-git related files to a sub directory and git-shell related to another sub directory if we cannot make all operations to an interface?

**Comment by zeripath** [comment]
@lunny that's the intended endpoint of this pr - once I've got the bugs ironed out and a more efficient getcommitpaths

**Comment by zeripath** [comment]
OK so the changes are slightly slower in GetLastCommitForPaths:

```
Nogogit: 2020/11/23 18:12:29 Completed GET /administrator/git 0  in 5.61394848s
GoGit: 2020/11/23 18:14:42 Completed GET /administrator/git 0  in 4.160055868s
```

This is predominantly because the algorithm here is different and simply involves checking every commit. This is clearly not going to be the fastest solution.

File: modules/repository/cache.go
Comment (lunny) [inline modules/repository/cache.go]: The first parameter could be ignored since commitCache should have chance to store it.

**Comment by zeripath** [comment]
damn the .drone.yml config is incorrect - hence the failing tests here.

**Comment by lunny** [comment]
Is this still wip?

**Comment by zeripath** [comment]
OK as per discussion on #maintainers I've changed this to default to using the native no-go-git variants. The go-git variants are possible to be built using the `gogit` build tag instead.

The sqlite and postgres tests will run using the go-git variants whilst the others will use the native variants.

I'm taking this out of WIP as I think we should consider adding this as it stands due to the issues with memory with go-git whilst I work on a runtime configurable solution in #13929.

File: Makefile
Comment (zeripath) [reply Makefile]: As the TEST_TAGS may be changed I think we should emit the test command.

File: Makefile
Comment (zeripath) [reply Makefile]: similarly

File: Makefile
Comment (zeripath) [reply Makefile]: similarly.

File: Makefile
Comment (6543) [reply Makefile]: @zeripath I still think this is to verbose

what about:
```diff
- @echo "Running go test..."
+ @echo "Running go test with '$(TEST_TAGS)' ..."
```
instead?

**Comment by 6543** [comment]
@zeripath can you add one integration test for gogit?

**Review by 6543 — APPROVED** [review]
as i can tell this is a huge imprvement

If I missed a bug I'm sorry - but I think we shoul get it in

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Let's merge this at first even if maybe there are some missed.

File: .drone.yml
Comment (silverwind) [inline .drone.yml:45]: Sorry I'm late here but is it really necessary to run the linter twice here? Do `TAGS` really affect the linting result? I thought `TAGS` only affect the build.

Comment (zeripath) [reply .drone.yml:45]: yes

Comment (silverwind) [reply .drone.yml:45]: I guess it's beyond me to understand how TAGS can affect linting (which should just be a static analysis of files). Care to explain? Does the go linter actually do runtime analysis as well?

Comment (lunny) [reply .drone.yml:45]: golangci-lint has a `--build-tags` to support lint different tags.

Comment (silverwind) [reply .drone.yml:45]: Ok I guess I need to study `golangci-lint` more. I was under the impression that linters in general just do static analysis of file contents, but I guess this linter does more than that.

Comment (silverwind) [reply .drone.yml:45]: Thought I generally wonder if it's really necessary to do this double linting every single CI run. The go linting already takes around 3 minutes per run.

Comment (zeripath) [reply .drone.yml:45]: The issue is that the go-git variants require linting too.

Comment (silverwind) [reply .drone.yml:45]: Can't we just pass a list of files to the linters as argument? I think it should be possible.

**Pull request merged**
